### PR TITLE
Fix Compiler Warnings

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
- Always treat warnings as error irrespective of debug string
- Fixes unused allocator variable in two places.
- No need to handle error `InitDefaultClient` as that can't fail.
- Handle error properly for TlsContextOptions.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by SergeyRyabinin — APPROVED** [review]

File: source/external/cJSON.cpp [hunk 1]
```diff
@@ -426,7 +426,7 @@ CJSON_PUBLIC(char*) cJSON_SetValuestring(cJSON *object, const char *valuestring)
     }
     if (strlen(valuestring) <= strlen(object->valuestring))
     {
-        strncpy(object->valuestring, valuestring, strlen(valuestring) + 1);
```
Comment (waahm7) [inline source/external/cJSON.cpp:429]: This causes 
```
error: char* strncpy(char*, const char*, size_t) specified bound depends on the length of the source argument [-Werror=stringop-truncation]
  429 |         strncpy(object->valuestring, valuestring, strlen(valuestring) + 1);
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/aws-crt-cpp/source/external/cJSON.cpp:427:15: note: length computed here
  427 |     if (strlen(valuestring) <= strlen(object->valuestring))
  ```

File: include/aws/crt/mqtt/Mqtt5Packets.h [hunk 2]
```diff
@@ -448,10 +448,7 @@ namespace Aws
             class AWS_CRT_CPP_API NegotiatedSettings
             {
               public:
-                NegotiatedSettings(
-                    const aws_mqtt5_negotiated_settings &negotiated_settings,
-
-                    Allocator *allocator = ApiAllocator()) noexcept;
+                NegotiatedSettings(const aws_mqtt5_negotiated_settings &negotiated_settings) noexcept;
```
Comment (graebm) [inline include/aws/crt/mqtt/Mqtt5Packets.h:452]: Don't change this. The pattern is that everything takes a custom allocator. This datastructure just happens to not need it at the moment

instead, go into the .cpp file and mark it unused via `(void)allocator;`

File: include/aws/crt/mqtt/Mqtt5Packets.h [hunk 3]
```diff
@@ -2209,9 +2206,7 @@ namespace Aws
             class AWS_CRT_CPP_API UnSubAckPacket : public IPacket
             {
               public:
-                UnSubAckPacket(
-                    const aws_mqtt5_packet_unsuback_view &packet,
-                    Allocator *allocator = ApiAllocator()) noexcept;
+                UnSubAckPacket(const aws_mqtt5_packet_unsuback_view &packet) noexcept;
```
Comment (graebm) [inline include/aws/crt/mqtt/Mqtt5Packets.h:2209]: same

File: source/external/cJSON.cpp (hunk 1 above)
Comment (TwistedTwigleg) [reply source/external/cJSON.cpp:429]: Question: Should this be `size_t value_length = strlen(valuestring) + sizeof("");` like in aws-c-common: https://github.com/awslabs/aws-c-common/blob/3973e3c89bbb532f5c3d2327b602836a57bd4db6/source/external/cJSON.c#L410?

File: CMakeLists.txt [hunk 4]
```diff
@@ -332,13 +332,13 @@ endif()
...
-    if(MSVC)
-        target_compile_options(${PROJECT_NAME} PRIVATE /W4 /WX /wd4068)
-    else()
-        target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wno-long-long -pedantic -Werror)
-    endif()
+ # set extra warning flags
+if(MSVC)
+    target_compile_options(${PROJECT_NAME} PRIVATE /W4 /WX /wd4068)
+else()
+    target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wno-long-long -pedantic -Werror)
```
[the remark this answers is not in the document]
Comment (waahm7) [reply CMakeLists.txt:341]: Thanks, makes sense.

File: source/external/cJSON.cpp (hunk 1 above)
Comment (waahm7) [reply source/external/cJSON.cpp:429]: I think `sizeof("")` is always 1. So it doesn't make a difference. https://stackoverflow.com/questions/4141666/why-sizeof-is-equivalent-to-1-and-sizeofnull-is-equivalent-to-4-in-c-langu

Comment (TwistedTwigleg) [reply source/external/cJSON.cpp:429]: > Not related, but should we remove the copy of cJSON from aws-crt-cpp, as we have one in aws-c-common already?

I am working on this now! It pretty much maps 1:1 with what we have in `aws-c-common`.

**Review by graebm — APPROVED** [review]

**Pull request merged**